django/db/models/expressions.py

```diff
@@ -547,6 +550,58 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None,
         return ResolvedOuterRef(self.name)
 
 
+class SliceableF(F):
+    """
+    An object that can add slicing notation to F expression.
+
+    Object resolves the column on which the slicing is applied, and then
+    applies the slicing if possible.
+    """
+    def __init__(self, f_obj, slice_obj):
+        if not isinstance(slice_obj, (int, slice)):
+            raise TypeError("Argument to slice must be either int or slice instance.")
+
+        is_negative_int = isinstance(slice_obj, int) and (slice_obj < 0)
+        has_negative_slice = (
+            isinstance(slice_obj, slice) and
+            ((slice_obj.start is not None and slice_obj.start < 0) or
+             (slice_obj.stop is not None and slice_obj.stop < 0))
+        )
+        if (is_negative_int or has_negative_slice):
+            raise ValueError("Negative indexing is not supported.")
+
+        if isinstance(slice_obj, slice) and slice_obj.step is not None:
+            raise ValueError("Step argument is not supported.")
+
+        self.expression = f_obj
+        if isinstance(slice_obj, slice):
+            if slice_obj.start is not None:
+                self.low = int(slice_obj.start) + 1
+            else:
+                self.low = 1
+
+            if slice_obj.stop is not None:
+                self.length = int(slice_obj.stop) - self.low + 1
+            else:
+                self.length = None
+        else:
+            self.low = int(slice_obj)
+            self.length = 1
```

Comment:
I think it will be more readable to keep `int` and `slice` in separate branches, e.g.:
```python
    def __init__(self, f_obj, slice_obj):
        if isinstance(slice_obj, int):
            if slice_obj < 0:
                raise ValueError('Negative indexing is not supported.')
            self.low = slice_obj
            self.length = 1
        elif isinstance(slice_obj, slice):
            if (
                (slice_obj.start is not None and slice_obj.start < 0) or
                (slice_obj.stop is not None and slice_obj.stop < 0)
            ):  
                raise ValueError('Negative indexing is not supported.')
            if slice_obj.step is not None:
                raise ValueError('Step argument is not supported.')
            self.low = 1 if slice_obj.start is None else int(slice_obj.start) + 1
            self.length = None if slice_obj.stop is None else int(slice_obj.stop) - self.low + 1
        else:
            raise TypeError('Argument to slice must be either int or slice instance.')
        self.expression = f_obj
```